### PR TITLE
[Feature] Add/Remove workers dynamically.

### PR DESCRIPTION
**Pull request by startswithaj**
### Description

I'm writing another project that monitors/scales firebase queues using wait-time and process-time and available system resources.

The way I currently have this working is to treat `Queues` as singleton workers and pass a factory method to my monitor/scaler which creates another `Queue` when it needs to add another worker. This is not logically optimal and adds unnecessary overhead (extra connections to firebase for the task and spec listeners). I could ofcourse simulataneous shutdown the current `Queue` and start another with my desired numWorkers param. This again seems unnecessary and a little dirty.

I think it would be better for me (and maybe others) if I could add/reduce the number of workers without having to A. create another `Queue`, B. Shutdown the current `Queue` start it again with new numWorkers parameter.

Huge N.B. caveat: I don't fully understand the `QueueWorker.prototype.setTaskSpec` and `QueueWorker.prototype._setUpTimeouts` methods. 

If @mbleigh or @drtriumph have time to look over this be much appreciated.

### Code sample

```
var q = new Queue(th.testRef, _.noop);
q.getWorkerCount()
q.addWorker()
q.removeWorker()
```

### Linter and test suite

Linting and tests passing

### Added tests

I added several tests. To cover the bulk of my changes.
The outstanding case is where a worker is added to a queue with a specified specID and before that spec has been retrieved from firebase. This is accounted for by not covered by tests.

I tried to write this with as few changes to the existing code. It would be good if things like creating the new queueWorker were refactored out and there weren't multiple for loops for setting the task specs of the workers.

### Sign our CLA
Signed

### CODE REVIEW

**Comment by startswithaj** [comment]
It seems with some incantation I have accidentally modified .travis.yml

https://github.com/firebase/firebase-queue/blob/master/.travis.yml

To point to my test fb... 

```
sudo: false
env:
  FB_QUEUE_TEST_DB_URL=https://fir-queue-test.firebaseio.com
install:
```

**Comment by startswithaj** [comment]
ello?

**Comment by abehaskins** [comment]
cc @drtriumph if you haven't seen this already

File: src/queue.js [hunk 1]
```diff
@@ -204,4 +204,50 @@ Queue.prototype.shutdown = function() {
...
+ */
+Queue.prototype.shutdownWorker = function() {
+  logger.debug('Queue: removing worker');
+  var shutdownPromise;
+  var worker = this.workers.shift();
+
+  if (worker) {
+    shutdownPromise = worker.shutdown();
+  }
+  return shutdownPromise;
```
Comment (drtriumph) [inline src/queue.js]: This should probably always return a Promise - if there are no more workers maybe it can return `RSVP.Resolve()`?

File: test/lib/queue_worker.spec.js [hunk 2]
```diff
@@ -2117,7 +2117,7 @@ describe('QueueWorker', function() {
         setTimeout(function() {
           callbackComplete = true;
           resolve();
-        }, 250);
+        }, 100);
```
Comment (drtriumph) [inline test/lib/queue_worker.spec.js]: Were these changes intentional?

File: src/queue.js [hunk 3]
```diff
@@ -204,4 +204,50 @@ Queue.prototype.shutdown = function() {
   }));
 };
 
+/**
+ * Gets queue worker count
```
Comment (drtriumph) [inline src/queue.js]: These need better JSDoc to describe the input/output

File: test/queue.spec.js [hunk 4]
```diff
@@ -144,6 +144,52 @@ describe('Queue', function() {
...
+    });
+  });
+
+  describe('#addWorker', function() {
+    it('should add worker', function() {
+      var q = new th.Queue(th.testRef, _.noop);
+      expect(q.getWorkerCount()).to.equal(1);
+      var worker = q.addWorker()
+      expect(q.getWorkerCount()).to.equal(2);
+    });
```
Comment (drtriumph) [inline test/queue.spec.js]: nit: need a new line between tests

File: test/queue.spec.js [hunk 5]
```diff
@@ -144,6 +144,52 @@ describe('Queue', function() {
...
+      expect(q.getWorkerCount()).to.equal(1);
+      var workerShutdownPromise = q.shutdownWorker()
+      return workerShutdownPromise
+    });
+    it('should return undefined when no workers remaining', function() {
+      var q = new th.Queue(th.testRef, _.noop);
+      expect(q.getWorkerCount()).to.equal(1);
+      q.shutdownWorker()
+      var workerShutdownPromise = q.shutdownWorker()
+      expect(workerShutdownPromise).to.be.undefined
```
Comment (drtriumph) [inline test/queue.spec.js]: This should be a Promise that resolves

File: src/queue.js [hunk 6]
```diff
@@ -204,4 +204,50 @@ Queue.prototype.shutdown = function() {
...
+    this.sanitize,
+    this.suppressStack,
+    this.processingFunction
+  );
+  this.workers.push(worker);
+
+  if (_.isUndefined(this.specId)) {
+    worker.setTaskSpec(DEFAULT_TASK_SPEC);
+  // if the currentTaskSpec is not yet set it will be called once it's fetched
+  } else if (this.currentTaskSpec) {
```
Comment (drtriumph) [inline src/queue.js]: this is inconsistent - could you make use of `_.isUndefined()`

File: src/queue.js [hunk 7]
```diff
@@ -204,4 +204,50 @@ Queue.prototype.shutdown = function() {
...
+  return worker;
+};
+
+/**
+ * Removes a queue worker if one exists
+ */
+Queue.prototype.shutdownWorker = function() {
+  logger.debug('Queue: removing worker');
+  var shutdownPromise;
+  var worker = this.workers.shift();
```
Comment (drtriumph) [inline src/queue.js]: Why are you using `shift()`? `pop()` is more efficient

File: src/queue.js [hunk 8]
```diff
@@ -204,4 +204,50 @@ Queue.prototype.shutdown = function() {
...
+
+  if (worker) {
```
Comment (drtriumph) [inline src/queue.js]: again, please use `_.isUndefined()`

File: src/queue.js (hunk 7 above)
Comment (drtriumph) [inline src/queue.js]: Also, as it stands, using a combination of `addWorker()` and `shutdownWorker()` will cause multiple workers to have the same processId prefix (though the UUID should never collide so it's still safe, just confusing if you're trying to debug)

File: test/queue.spec.js [hunk 9]
```diff
@@ -144,6 +144,52 @@ describe('Queue', function() {
...
+    });
+  });
+
+  describe('#shutdownWorker', function() {
+    it('should remove worker', function() {
+      var q = new th.Queue(th.testRef, _.noop);
+      expect(q.getWorkerCount()).to.equal(1);
+      var workerShutdownPromise = q.shutdownWorker()
+      expect(q.getWorkerCount()).to.equal(0);
+    });
```
Comment (drtriumph) [inline test/queue.spec.js]: nit: need new line below

File: test/queue.spec.js [hunk 10]
```diff
@@ -144,6 +144,52 @@ describe('Queue', function() {
...
+    it('should shutdown worker', function() {
+      var q = new th.Queue(th.testRef, _.noop);
+      expect(q.getWorkerCount()).to.equal(1);
+      var workerShutdownPromise = q.shutdownWorker()
+      return workerShutdownPromise
+    });
```
Comment (drtriumph) [inline test/queue.spec.js]: nit: need new line below

**Comment by drtriumph** [comment]
Alright, first pass through this. It looks fairly simple and I'd be happy to merge it in if you address all the issues. Thanks for the taking the time to share

File: src/queue.js (hunk 1 above)
Comment (startswithaj) [inline src/queue.js]: Yeap no worries. My intention was to provider a way to tell if a worker was shutdown or not from the call, rather than query totalWorkers before and after. I actually don't need this functionality in my use case so I'm happy to always resolve.

A couple of things I considered:
a. throwing an error if there was no more workers to shutdown.
b. returning a promise that resolved when the shutdown finish with the worker or undefined if there was no more workers to shutdown
c. as you suggested returning a resolved promise

I'm leaning towards a. What do you think @drtriumph

File: src/queue.js (hunk 7 above)
Comment (startswithaj) [inline src/queue.js]: Shift vs Pop: I'll change it to pop. I think superstitiously I thought it might be better to shutdown the oldest worker. 

ID: True point. Would you like me to add a counter?

File: src/queue.js (hunk 8 above)
Comment (startswithaj) [inline src/queue.js]: No problem.

File: test/lib/queue_worker.spec.js (hunk 2 above)
Comment (startswithaj) [inline test/lib/queue_worker.spec.js]: 2120: No

2150: Yes (On my workstation expect(callbackStarted).to.be.true fails because the queue hasn't picked up the task yet.

I will fix 2120 it was an artifact of fiddling around with it.

File: test/queue.spec.js (hunk 5 above)
Comment (startswithaj) [inline test/queue.spec.js]: roger that.

File: test/queue.spec.js (hunk 10 above)
Comment (startswithaj) [inline test/queue.spec.js]: roger all ^^^

File: src/queue.js (hunk 1 above)
Comment (drtriumph) [inline src/queue.js]: d. return `RSVP.reject(new Error('No workers to shutdown'))` or similar

I think it should be a promise rather than `undefined` or a synchronous error, and I can see an argument that this is a failure case - I'm leaning towards d.

File: src/queue.js (hunk 7 above)
Comment (drtriumph) [inline src/queue.js]: I can see the logic there.

If you use `pop()` then you don't have to worry about the counter - the way you have things will ensure the workers always get their current index and you'll never have two workers at any point in time with the same process ID prefix. I don't think we need a counter to ensure they're unique for all time the process is running

File: src/queue.js (hunk 6 above)
Comment (startswithaj) [inline src/queue.js]: _.isUndefined makes sense for things that can be falsy when set. Like this.specId could equal 0.

But this.currentTaskSpec will be an object truthy or nothing. I would inclined to argue !_.isUndefined(this.currentTaskSpec) is far less readable than if (this.currentTaskSpec).

Apart from consistency and protecting against falsy values is there a reason to use _.isUndefined.

I will change it, but I'm genuinely interested.

**Comment by startswithaj** [comment]
@drtriumph thanks for looking at this for me. With my latest 2 commits I think I have addressed all the stuff you listed above:

- isUndefined consistency
- new line consistency
- shutdown always returns promise (reject on no workers to shutdown)
- better JSDoc for new functions
- revert incorrect test change

File: src/queue.js (hunk 6 above)
Comment (drtriumph) [inline src/queue.js]: It's not always possible to know if something might return falsy in such a loosely-typed language - in this case we don't have a lot of control over what the external dependency returns. Though it's very unlikely that the Firebase SDK will return anything that evaluates falsy for a reference, it doesn't hurt to be explicit about precisely what we want to test for (if they did, for example, that shouldn't change the behavior here).

However, even if the value were completely under our control, I think consistency should trump efficiency

**Comment by drtriumph** [comment]
Closing in favor of #72
